Improve Get/Put performance with optional mempooling

jsign opened the pull request:
A continued work of #32

Changes on top of what the original PR improved:
* Using a mempool for avoid 1 allocation in Get.
* Configuration of mempooling is optional. The default behavior is not using mempool.
* Additional table-tests with mempool to compare face-to-face.
* Rollback of the mempool proposed in the `streampb` package.

I went forward and backward with the type of pool to use. My original intention was `SizedBufferPool` but that was quite complicated because the max size of each element in the pool would be the maximum size of `proto.Entry`, which isn't trivial to calculate considering `maxKeySize` and `maxValueSize`.

As a solution, I proposed using `BufferPool` and always see if the current `Buffer` is enough to the requested size. If isn't the case, Grow it appropriately. The pool entries will eventually converge to the worst usual case.

I also considered adding one more parameter to the new `Option`: `MaxTotalMemPoolSize` to set a hard-limit on mempool size, but I felt wasn't quite trivial to use. Each max buff in the pool would be `MaxTotalMemPoolSize/MaxConcurrency`, and if `MaxTotalMemPoolSize` is set too low would be a problem. Moreover, estimating a reasonable value implies trying to guess what is the expected encoded proto size for values. (This may be confusing, I can reword it if isn't that clear).

Results:
```
goos: linux
goarch: amd64
pkg: github.com/prologic/bitcask
BenchmarkGet/128B-8              1000000              1152 ns/op         111.02 MB/s         387 B/op          4 allocs/op
BenchmarkGet/128BWithPool-8      1000000              1189 ns/op         107.59 MB/s         227 B/op          3 allocs/op
BenchmarkGet/256B-8              1000000              1255 ns/op         203.86 MB/s         643 B/op          4 allocs/op
BenchmarkGet/256BWithPool-8      1000000              1246 ns/op         205.33 MB/s         355 B/op          3 allocs/op
BenchmarkGet/512B-8              1000000              1416 ns/op         361.47 MB/s        1187 B/op          4 allocs/op
BenchmarkGet/512BWithPool-8      1000000              1357 ns/op         377.10 MB/s         611 B/op          3 allocs/op
BenchmarkGet/1K-8                1000000              1763 ns/op         580.76 MB/s        2275 B/op          4 allocs/op
BenchmarkGet/1KWithPool-8        1000000              1592 ns/op         642.89 MB/s        1123 B/op          3 allocs/op
BenchmarkGet/2K-8                 500000              2415 ns/op         847.88 MB/s        4451 B/op          4 allocs/op
BenchmarkGet/2KWithPool-8        1000000              2009 ns/op        1019.35 MB/s        2147 B/op          3 allocs/op
BenchmarkGet/4K-8                 300000              3854 ns/op        1062.67 MB/s        9059 B/op          4 allocs/op
BenchmarkGet/4KWithPool-8         500000              3029 ns/op        1352.22 MB/s        4195 B/op          3 allocs/op
BenchmarkGet/8K-8                 200000              6173 ns/op        1326.90 MB/s       17763 B/op          4 allocs/op
BenchmarkGet/8KWithPool-8         300000              4911 ns/op        1668.07 MB/s        8291 B/op          3 allocs/op
BenchmarkGet/16K-8                200000             10697 ns/op        1531.57 MB/s       34915 B/op          4 allocs/op
BenchmarkGet/16KWithPool-8        200000              7763 ns/op        2110.51 MB/s       16483 B/op          3 allocs/op
BenchmarkGet/32K-8                100000             21858 ns/op        1499.11 MB/s       73827 B/op          4 allocs/op
BenchmarkGet/32KWithPool-8        100000             14607 ns/op        2243.25 MB/s       32867 B/op          3 allocs/op
```
Changes in the PR branch are allowed, so feel free to add further commits.

prologic commented:
Here are the same results on my 11" Macbook:

```
--- PASS: TestLocking (0.01s)
goos: darwin
goarch: amd64
pkg: github.com/prologic/bitcask
BenchmarkGet/128B-4         	  300000	      4126 ns/op	  31.02 MB/s	     387 B/op	       4 allocs/op
BenchmarkGet/128BWithPool-4 	  300000	      4275 ns/op	  29.94 MB/s	     227 B/op	       3 allocs/op
BenchmarkGet/256B-4         	  300000	      3970 ns/op	  64.47 MB/s	     643 B/op	       4 allocs/op
BenchmarkGet/256BWithPool-4 	  300000	      4570 ns/op	  56.01 MB/s	     355 B/op	       3 allocs/op
BenchmarkGet/512B-4         	  300000	      4843 ns/op	 105.71 MB/s	    1187 B/op	       4 allocs/op
BenchmarkGet/512BWithPool-4 	  200000	      5622 ns/op	  91.07 MB/s	     611 B/op	       3 allocs/op
BenchmarkGet/1K-4           	  300000	      5885 ns/op	 173.99 MB/s	    2275 B/op	       4 allocs/op
BenchmarkGet/1KWithPool-4   	  200000	      5886 ns/op	 173.96 MB/s	    1123 B/op	       3 allocs/op
BenchmarkGet/2K-4           	  200000	      6859 ns/op	 298.56 MB/s	    4451 B/op	       4 allocs/op
BenchmarkGet/2KWithPool-4   	  200000	      6724 ns/op	 304.55 MB/s	    2147 B/op	       3 allocs/op
BenchmarkGet/4K-4           	  200000	      8271 ns/op	 495.21 MB/s	    9059 B/op	       4 allocs/op
BenchmarkGet/4KWithPool-4   	  200000	      7429 ns/op	 551.32 MB/s	    4195 B/op	       3 allocs/op
BenchmarkGet/8K-4           	  200000	     10900 ns/op	 751.53 MB/s	   17763 B/op	       4 allocs/op
BenchmarkGet/8KWithPool-4   	  200000	      9332 ns/op	 877.77 MB/s	    8291 B/op	       3 allocs/op
BenchmarkGet/16K-4          	  100000	     14014 ns/op	1169.07 MB/s	   34915 B/op	       4 allocs/op
BenchmarkGet/16KWithPool-4  	  100000	     12152 ns/op	1348.19 MB/s	   16483 B/op	       3 allocs/op
BenchmarkGet/32K-4          	  100000	     23211 ns/op	1411.70 MB/s	   73827 B/op	       4 allocs/op
BenchmarkGet/32KWithPool-4  	  100000	     18618 ns/op	1759.97 MB/s	   32867 B/op	       3 allocs/op
BenchmarkPut/128B-4         	  100000	     13048 ns/op	   9.81 MB/s	     409 B/op	       6 allocs/op
BenchmarkPut/256B-4         	  100000	     15815 ns/op	  16.19 MB/s	     539 B/op	       6 allocs/op
BenchmarkPut/512B-4         	  100000	     20986 ns/op	  24.40 MB/s	     829 B/op	       6 allocs/op
BenchmarkPut/1K-4           	   50000	     27889 ns/op	  36.72 MB/s	    1411 B/op	       6 allocs/op
BenchmarkPut/2K-4           	   50000	     37723 ns/op	  54.29 MB/s	    2574 B/op	       6 allocs/op
BenchmarkPut/4K-4           	   20000	     72584 ns/op	  56.43 MB/s	    5156 B/op	       6 allocs/op
BenchmarkPut/8K-4           	   10000	    124705 ns/op	  65.69 MB/s	    9814 B/op	       6 allocs/op
BenchmarkPut/16K-4          	   10000	    195204 ns/op	  83.93 MB/s	   18851 B/op	       6 allocs/op
BenchmarkPut/32K-4          	    5000	    350503 ns/op	  93.49 MB/s	   41549 B/op	       7 allocs/op
BenchmarkScan-4             	 1000000	      1943 ns/op	     493 B/op	      25 allocs/op
PASS
ok  	github.com/prologic/bitcask	49.615s
```

prologic commented:
Are we also merging what's in #32 in addition to this PR? Sorry I'm a little confused :) I thought I was just going to drop #32 in favor of this :)

jsign commented:
This PR in a fork of your branch in the other PR. This PR has the same commits you authored in the other PR. Since I'm not a contributor, I couldn't push new commits to your branch to continue on the other PR.

I rebased to master too.

prologic commented:
Oh nevermind :) I just woke up and haven't had enough caffeine! My commits are in your PR!

prologic commented:
> Since I'm not a contributor, I couldn't push new commits to your branch to continue on the other PR (or thats what I think since I coudn't push to it)

Oh that can easily be fixed! One moment...

prologic commented:
Please feel free to push to direct branches. I have a few minor comments on this PR:

- Why do we not just set `memPool` to `nil` by default and avoid passing in int poiners `*int`?
- The boop library calls the paramter `maxSize` and we're callig it `maxConcurrency`? Should we stick with `maxSize` to be consistent?

prologic commented:
Should we also employ the use of a fixed-size `ByteBuffer` for allocating the prefixSize framing used which will just speed up reading/streaming the whole database (_in merge operations_)?

```
prologic@Jamess-MacBook
Sun Aug 04 09:39:51
~/Projects/bitcask
 (buffer_pool2) 0
$ sift -C 3 'make.*byte'
internal/datafile.go-143-
internal/datafile.go-144-	var b []byte
internal/datafile.go-145-	if memPool == nil {
internal/datafile.go:146:		b = make([]byte, size)
internal/datafile.go-147-	} else {
internal/datafile.go-148-		poolSlice := memPool.Get()
internal/datafile.go-149-		if poolSlice.Cap() < int(size) {
--
internal/streampb/stream.go-29-// Encode takes any proto.Message and streams it to the underlying writer.
internal/streampb/stream.go-30-// Messages are framed with a length prefix.
internal/streampb/stream.go-31-func (e *Encoder) Encode(msg proto.Message) (int64, error) {
internal/streampb/stream.go:32:	prefixBuf := make([]byte, prefixSize)
internal/streampb/stream.go-33-
internal/streampb/stream.go-34-	buf, err := proto.Marshal(msg)
internal/streampb/stream.go-35-	if err != nil {
--
internal/streampb/stream.go-67-// Decode takes a proto.Message and unmarshals the next payload in the
internal/streampb/stream.go-68-// underlying io.Reader. It returns an EOF when it's done.
internal/streampb/stream.go-69-func (d *Decoder) Decode(v proto.Message) (int64, error) {
internal/streampb/stream.go:70:	prefixBuf := make([]byte, prefixSize)
internal/streampb/stream.go-71-
internal/streampb/stream.go-72-	_, err := io.ReadFull(d.r, prefixBuf)
internal/streampb/stream.go-73-	if err != nil {
--
internal/streampb/stream.go-76-
internal/streampb/stream.go-77-	n := binary.BigEndian.Uint64(prefixBuf)
internal/streampb/stream.go-78-
internal/streampb/stream.go:79:	buf := make([]byte, n)
internal/streampb/stream.go-80-
internal/streampb/stream.go-81-	idx := uint64(0)
internal/streampb/stream.go-82-	for idx < n {
```

jsign commented:
> Please feel free to push to direct branches. I have a few minor comments on this PR:
> 
> * Why do we not just set `memPool` to `nil` by default and avoid passing in int poiners `*int`?
> * The boop library calls the paramter `maxSize` and we're callig it `maxConcurrency`? Should we stick with `maxSize` to be consistent?

* Since the `memPool` is package-wise, if a client creates a mempooled enabled `db`, close its, and then opens it again but want it without pool... this would allow to re-configure the package. This is similar to what I mentioned in the other PR regarding some refactoring. (I don't like package scoped variables for this kind of consequences)
* Maybe, but I think that's polluting the language domain of what the client understands. I guess the `maxSize` concept of `bpool` is an implementation detail which affects the concurrency concept in bitcask. It may be renamed as `maxSize`, but we should clarify a bit more _size of what_ , _why it matters for you_, and _how you should choose this value_ (naming `maxConcurrency` may imply the clarifications, but I guess gives a better hint; not sure about this though).

jsign commented:
> Should we also employ the use of a fixed-size `ByteBuffer` for allocating the prefixSize framing used which will just speed up reading/streaming the whole database (_in merge operations_)?
> 
> ```
> prologic@Jamess-MacBook
> Sun Aug 04 09:39:51
> ~/Projects/bitcask
>  (buffer_pool2) 0
> $ sift -C 3 'make.*byte'
> internal/datafile.go-143-
> internal/datafile.go-144-	var b []byte
> internal/datafile.go-145-	if memPool == nil {
> internal/datafile.go:146:		b = make([]byte, size)
> internal/datafile.go-147-	} else {
> internal/datafile.go-148-		poolSlice := memPool.Get()
> internal/datafile.go-149-		if poolSlice.Cap() < int(size) {
> --
> internal/streampb/stream.go-29-// Encode takes any proto.Message and streams it to the underlying writer.
> internal/streampb/stream.go-30-// Messages are framed with a length prefix.
> internal/streampb/stream.go-31-func (e *Encoder) Encode(msg proto.Message) (int64, error) {
> internal/streampb/stream.go:32:	prefixBuf := make([]byte, prefixSize)
> internal/streampb/stream.go-33-
> internal/streampb/stream.go-34-	buf, err := proto.Marshal(msg)
> internal/streampb/stream.go-35-	if err != nil {
> --
> internal/streampb/stream.go-67-// Decode takes a proto.Message and unmarshals the next payload in the
> internal/streampb/stream.go-68-// underlying io.Reader. It returns an EOF when it's done.
> internal/streampb/stream.go-69-func (d *Decoder) Decode(v proto.Message) (int64, error) {
> internal/streampb/stream.go:70:	prefixBuf := make([]byte, prefixSize)
> internal/streampb/stream.go-71-
> internal/streampb/stream.go-72-	_, err := io.ReadFull(d.r, prefixBuf)
> internal/streampb/stream.go-73-	if err != nil {
> --
> internal/streampb/stream.go-76-
> internal/streampb/stream.go-77-	n := binary.BigEndian.Uint64(prefixBuf)
> internal/streampb/stream.go-78-
> internal/streampb/stream.go:79:	buf := make([]byte, n)
> internal/streampb/stream.go-80-
> internal/streampb/stream.go-81-	idx := uint64(0)
> internal/streampb/stream.go-82-	for idx < n {
> ```
TBH, I think the best thing is that they could share the same pool... but this is further refactoring since they are in different packages (similar to what I mentioned above).
One option would be doing the refactoring, or other is applying the same configurable idea to avoid hardcoding concurrency. What do you think?

prologic commented:
Okay 👍Before we merge this; can we add some user docs to a `doc.go` so it show sup in godoc with some verbage about general use and the optional pooling?

Can we also clean up the commits a bit? Maybe feel free to push to a new branch and reword my commit? :) I sent you an invite to collaborate directly on the repo.

jsign commented:
> Approved with comments. Docs + Cleanup and we're good :)
Sure, I'll do the docs and rebase with a nicer commit history and force push here.

jsign commented:
* added small documentation in the package mentioning, in particular, the mempool feature.
* added examples in the documentation to show quickly how to use variadic configuration
* made a nicer commit history

prologic commented:
LGTM I'll merge this shortly after CI runs

jsign commented:
Great. I think may be a good idea later to update the Performance section in the Readme to show these improvements. Maybe using the same computer that lead the the current results to make a fair history of the Readme.

prologic commented:
Sure I’ll do that today
-- 

James Mills / prologic

E: prologic@shortcircuit.net.au
W: prologic.shortcircuit.net.au
